File: LeetCode/src/binarytree/MaximumDepthOfBinaryTree.java
package binarytree;

import java.util.LinkedList;
import java.util.Queue;

public class MaximumDepthOfBinaryTree {
    public int maxDepth(TreeNode root) {
//        return dfs(root);
        return bfs(root);
    }

    private int bfs(TreeNode node) {
        if (node == null) {
            return 0;
        }

        Queue<TreeNode> queue = new LinkedList<>();
        queue.offer(node);

        int result = 0;

        while (!queue.isEmpty()) {
            int size = queue.size();

            for (int i = 0; i < size; i++) {
                TreeNode n = queue.poll();

                if (n.left != null) queue.offer(n.left);
                if (n.right != null) queue.offer(n.right);
            }

            result++;
        }

        return result;
    }

    private int dfs(TreeNode node) {
        if (node == null) {
            return 0;
        }

        return Math.max(maxDepth(node.left), maxDepth(node.right)) + 1;
    }

    public static void main(String[] args) {
        MaximumDepthOfBinaryTree solution = new MaximumDepthOfBinaryTree();

        // Test case
        TreeNode root = new TreeNode(3);
        root.left = new TreeNode(9);
        root.right = new TreeNode(20);
        root.right.left = new TreeNode(15);
        root.right.right = new TreeNode(7);

        System.out.println("Maximum Depth of Binary Tree: " + solution.maxDepth(root)); // Output: 3
    }
}

File: LeetCode_Python/binarytree/MaximumDepthOfBinaryTree.py
from TreeNode import TreeNode
from typing import Optional


class MaximumDepthOfBinaryTree:
    def maxDepth(self, root: Optional[TreeNode]) -> int:
        def dfs(node: TreeNode) -> int:
            if not node:
                return 0

            left_depth = dfs(node.left)
            right_depth = dfs(node.right)

            return max(left_depth, right_depth) + 1

        return dfs(root)

    def maxDepthByBFS(self, root: Optional[TreeNode]) -> int:
        if not root:
            return 0

        queue = [root]
        depth = 0

        while queue:
            depth += 1
            for _ in range(len(queue)):
                node = queue.pop(0)
                if node.left:
                    queue.append(node.left)
                if node.right:
                    queue.append(node.right)

        return depth


# Time Complexity: O(n) where n is the number of nodes in the binary tree
# Space Complexity: O(h) where h is the height of the binary tree due to recursion stack space


if __name__ == "__main__":
    solution = MaximumDepthOfBinaryTree()

    root1 = TreeNode(3)
    root1.left = TreeNode(9)
    root1.right = TreeNode(20)
    root1.right.left = TreeNode(15)
    root1.right.right = TreeNode(7)
    print("Expected Output: 3")
    print("Actual Output: ", solution.maxDepth(root1))
    print("Actual Output: ", solution.maxDepthByBFS(root1))

    root2 = TreeNode(1)
    root2.left = TreeNode(2)
    print("Expected Output: 2")
    print("Actual Output: ", solution.maxDepth(root2))
    print("Actual Output: ", solution.maxDepthByBFS(root2))
